Good-Bids/goodbids
# Extending and Closing Auctions

# Summary

This PR extends the Auction "Close Date/Time" value whenever a bid is placed within the bid window. The bid window is determined by the Current Auction Close Date/Time minus the Bid Extension Value. When an Auction reaches its closing Date/Time, it signals to the Auctioneer and closes the Auction.

### Other Stuff in this PR

* Adds new Bid Extension Auction field.
* Disable/Enable Bid Now button based on Auction status.
* Checkout Validation when placing a bid - Confirms Auction is active
* Automatically empty the cart before adding a Bid or Reward Product
* New Auctioneer Submodule for Auctions endpoint
* Small Auctioneer refactor and `$last_response` property for reference
* Fixes related to Date/Time calculations/comparisons
* Adjustments to Auction Settings Meta Box (ACF Fields)
* Slight tweak to Pull Request Template

## Issues

* #17 
* #33 
* #32 
* #88

## Testing Instructions

* [Prerequisite] Configure your [local environment variables](https://github.com/Good-Bids/goodbids/pull/116). (cc/@shascher for Documentation) - See Testing Instructions-ish > Local Environment
* [Prerequisite] Set up WooCommerce Stripe
* [Prerequisite] Create a Reward Product
* [Prerequisite] Make sure the site used for testing has the same Time Zone you are in (under Settings > General)

### Bid Extension

1. Create a new Auction
    1. Fill out all the required fields. (adjust numbers as needed to give you time to test)
    2. Schedule the Auction Start Date/Time to be 5min from the current time.
    3. Schedule the Auction End Date/Time to be 20min from the current time.
    4. Set the Bid Extension to 5min
    5. Publish the Auction
2. (Reload the Admin page to update the Auction Info Meta Box)
    * You should see the Status of "Upcoming" and how long until it starts.
3. Visit the Auction (before it starts)
    * You should see a message "Auction is currently not active."
4. When the Auction starts
    * You should see the Status of "Live"
    * You should see the Bid $ Now button.
5. Place a bid BEFORE the Auction reaches the extension window
    * You should notice under the Auction Metrics the Auction Close Date/Time should match the original Auction Close Date/Time.
6. AFTER the Bid Extension window has opened:
    * You should see the Status "Live (Extension Window)".
7. Place a bid AFTER the Bid Extension window has opened
    * You should notice the Auction Close Date/Time adjust - It should ADD the Bid Extension time to the Date/Time of WHEN THE BID WAS PLACED.
    * You should see the Status "Live (Extended)"
    * Total Extensions should increase by 1.
8. Click the Bid Now BEFORE the Auction Closes, however, do not Checkout until AFTER the Auction is Closed
    * When clicking the Place Order button, you should get an error "Auction has already ended."

## Screenshots
**Bid Now Button:** Pre-Auction Start
<img width="403" alt="Screenshot 2023-12-27 at 11 28 36 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/4ff8a8bf-793a-4c19-bf51-d3713e14081d">

**Auction Info Panel:** Pre-Auction Start
<img width="279" alt="Screenshot 2023-12-27 at 11 22 17 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/acfff9e8-f222-456b-8976-996202e2bbbc">

**Auction Info Panel:** Auction Live
<img width="278" alt="Screenshot 2023-12-27 at 11 31 11 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/bbd2f680-3ac8-4d6b-8498-581a25ea5d3b">

**Bid Now Button:** Auction Live
<img width="403" alt="Screenshot 2023-12-27 at 11 31 42 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/ef8ec178-b8c8-482f-8e15-51070ba8a6b7">

**Auction Info Panel:** Auction Extension Window
<img width="276" alt="Screenshot 2023-12-27 at 11 40 42 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/65e44510-858a-4adb-be04-451b84e064b8">

**Auction Info Panel:** Auction Extended
<img width="277" alt="Screenshot 2023-12-27 at 11 41 50 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/4e7d35c0-70bb-4487-ad0c-993871bdfb85">

**Auction Info Panel:** Auction Closed
<img width="277" alt="Screenshot 2023-12-27 at 11 52 11 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/96959991-4b78-4cfc-9a06-234086a599ff">

**Bid Now Button:** Auction Closed
<img width="428" alt="Screenshot 2023-12-27 at 11 52 39 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/f6eb6f82-e427-4e27-a88b-d66a9d9429a5">

**Auction Closed Bid Error**
<img width="1130" alt="Screenshot 2023-12-27 at 11 53 10 AM" src="https://github.com/Good-Bids/goodbids/assets/122309362/69538368-ec36-4643-8f64-d8fcc2f6a153">
